fendermint/testing/materializer/src/validation.rs

```diff
@@ -0,0 +1,382 @@
+        validators: BTreeMap<&'a VAccount, Collateral>,
+        balances: BTreeMap<&'a VAccount, Balance>,
+    ) -> anyhow::Result<VGenesis> {
+        self.ensure_contains(subnet_name)?;
+        let tn = self.network()?;
+
+        if validators.is_empty() {
+            bail!("validators of {subnet_name:?} cannot be empty");
+        }
+
+        let root_balances = self.balances.entry(tn.root()).or_default();
+
+        for (n, b) in balances {
+            let balance = root_balances.entry(n.clone()).or_default();
+            *balance = b.0;
+        }
+
+        Ok(())
+    }
+
+    async fn create_node<'s, 'a>(
+        &'s mut self,
+        node_name: &NodeName,
+        _node_config: NodeConfig<'a, ValidationMaterials>,
+    ) -> anyhow::Result<VNode>
+    where
+        's: 'a,
+    {
+        self.ensure_contains(node_name)?;
+        Ok(node_name.clone())
+    }
+
+    async fn start_node<'s, 'a>(
+        &'s mut self,
+        _node: &'a VNode,
+        _seed_nodes: &'a [&'a VNode],
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    async fn create_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        subnet_name: &SubnetName,
+        subnet_config: SubnetConfig<'a, ValidationMaterials>,
+    ) -> anyhow::Result<VSubnet>
+    where
+        's: 'a,
+    {
+        self.ensure_contains(subnet_name)?;
+        // Check that the submitter has balance on the parent subnet to create the child.
+        let parent = parent_name(subnet_name)?;
+        self.ensure_balance(&parent, subnet_config.creator)?;
+        // Insert child subnet balances entry.
+        self.balances
+            .insert(subnet_name.clone(), Default::default());
+        Ok(subnet_name.clone())
+    }
+
+    async fn fund_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        account: &'a VAccount,
+        subnet: &'a VSubnet,
+        amount: TokenAmount,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()>
+    where
+        's: 'a,
+    {
+        // Debit parent balance; Credit child balance
+        self.fund_from_parent(subnet, account, amount, true)?;
+        self.ensure_unique(&subnet.parent().unwrap(), reference)?;
+        Ok(())
+    }
+
+    async fn join_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        account: &'a VAccount,
+        subnet: &'a VSubnet,
+        collateral: Collateral,
+        balance: Balance,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()>
+    where
+        's: 'a,
+    {
+        // Debit parent balance, but do not make the funds available in the child
+        self.fund_from_parent(subnet, account, collateral.0, false)?;
+        // Debit parent balance; Credit child balance
+        self.fund_from_parent(subnet, account, balance.0, true)?;
```

Comment:
Doesn't this need to be atomic? Like if first one succeeds and second fails

Reply:
This is just validation, so I don't think it matters much. Originally I thought I'll just collect errors in a list and return all the problems at the end, rather than shortcut at the first, then you fix it and it dies on the second. I thought the [Validated](https://typelevel.org/cats/datatypes/validated.html) monad worked like that, or maybe it was [this older one](https://eed3si9n.com/learning-scalaz/Validation.html). 

Then I thought I'm not going to bother, the testnet can fail early on other problems too, and with the logs it's easier to see the steps end right where the error happened. 

So, this is really just a way to say "hold on, something's not right HERE" and let you figure it out, rather than a proper accounting system. 

On a "real" materializer, the `reference` should help to make things resumable, ie. if the first succeeded, the second failed, and you run the whole thing again, it should skip the first and only try the second.

Reply:
makes sense!